assets/bower_components/three-js/three-js.html

```diff
@@ -1418,3 +1418,49 @@
     });
   </script>
 </dom-module>
+
+<dom-module id="three-js-raycaster">
+  <script>
+    Polymer({
+      is: 'three-js-raycaster',
+      properties: {
+        signal: {
+          type: Number
+        },
+        event: {
+          type: String
+        }
+      },
+      newRaycaster: function () {
+        this.object = new THREE.Raycaster();
+
+        var renderer = this.fire('three-js-get-renderer').detail.renderer;
+        var scene = Polymer.dom(this).parentNode.objectParent.scene;
+        var camera = Polymer.dom(this).parentNode.objectParent.camera;
+
+        if (this.event) {
+            var that = this;
+            renderer.domElement.addEventListener(this.event, function (event) {
+                var boundingBox = renderer.domElement.getBoundingClientRect();
+
+                var x = 2 * ((event.clientX - boundingBox.left) / renderer.domElement.clientWidth) - 1;
+                var y = 2 * ((boundingBox.top - event.clientY) / renderer.domElement.clientHeight) + 1;
+                that.object.setFromCamera(new THREE.Vector2(x, y), camera);
+
+                var intersects = that.object.intersectObjects(scene.children);
+                if (intersects.length > 0) {
+                    Escher.send({command: 'signal-update', data: {signalId: that.signal, value: intersects[0].point}});
```

Comment:
Okay, so what you're trying to add would be called a "behavior" in Escher. (Note that it's not the same as behavior in polymer - Escher used the word before Polymer 1.0 first did), a `Behavior` is a subtype of `Tile`, but it can "give out a signal".

Let's consider the clickable behavior -- it takes a tile and makes it "clickable" - you can then listen to the click on the input tile. For example.
```julia
el=plaintext("Hello, World!")
clicks = Signal{Escher.MouseButton}(leftbutton)
el_subscribed = subscribe(clicks, clickable(el, buttons=[leftbutton, rightbutton]))
```

Now clicking on `el_subscribed` will update the `clicks` signal.

Here's how this works:

1) The output is something like this: 
```
    <div>Hello, World!
        <clickable-behavior buttons=[1, 3]>
           <signal-transport .../>
         </clickable-behavior>
</div>
```
So you can imagine that `clickable` added the `<clickable-behavior>` tag to the plain text, and `subscribe` added `signal-transport` to _that_ tag. See their respective "render" functions ([clickable](https://github.com/shashi/Escher.jl/blob/master/src/basics/behavior.jl#L112-L118) & [subscribe](https://github.com/shashi/Escher.jl/blob/master/src/basics/signal.jl#L339-L346).

2) When <clickable-behavior> element is loaded, it [adds a "click" listener](https://github.com/shashi/Escher.jl/blob/master/assets/behavior.html#L82-L90) to the parent element, in this handler, it detects if the left or the right mouse button was fired, then calls [`Mixins.ReactiveSignal.update`](https://github.com/shashi/Escher.jl/blob/master/assets/signals.html#L51) with `1` or `3` indicating left or right button - (2 is scroll button)

3) as you  probably noticed, the `Subscription` tile type has an`intent` field which is used together with the signal object. This is an object that tells Julia how to interpret updates coming in from the tile being subscribed to - in our case the `Clickable` tile. When an intent is not specified in `subscribe`, [a default intent is used](https://github.com/shashi/Escher.jl/blob/master/src/basics/signal.jl#L307-L308) (simply the return value of `default_intent` on the input tile). The default intent of a `Clickable` is `ClickIntent`. You may have also noticed that the numbers `1` and `3` are converted to `leftbutton` ==LeftButton() and `rightbutton` ==RightButton() objects. This is the purpose of the `ClickIntent`.

When a message is received from a behavior tile such as clickable-behavior, Escher calls the `interpret` function with the intent of the behavior and the message. In our case `interpret(ClickIntent(), 1)` or `interpret(ClickIntent(),2)` which [calls this](https://github.com/shashi/Escher.jl/blob/master/src/basics/behavior.jl#L105-L110) which returns the value that finally gets pushed to the signal that subscribed to the behavior.

4) There is one last optional piece to this dance: `wrapbehavior`. This is used to associate default behaviors to _Tiles_, for example, a [Button has a default behavior of `clickable`](https://github.com/shashi/Escher.jl/blob/master/src/library/widgets.jl#L60-L61). This makes it possible to say, `subscribe(clicks, button())` which becomes `subscribe(clicks, wrapbehavior(button()))` (but not `subscribe(clicks, "hi")` since "hi" doesn't have a wrapbehavior method.) I don't think wrapbehavior is useful for you.

Coming to this PR:

1. You will need to define a type `immutable Raycaster <: Behavior; options; tile end` which has all the configuration options you need. Then add a method to Escher.render to render it, something like `render(r::Raycaster, state) = render(r.tile, state) << Elem(:three-js-raycaster)`
2. Add the `EscherMixins.ReactiveSignal` polymer behavior to the three js raycaster element, then call `elem = this.findElem(); this.update(elem, {my data...})`
3. Create `RaycasterIntent` which can read `{my data...}` coming from JS to create Julia types representing the update. Also, you will need to define, `Escher.default_intent(x::Raycaster) = RaycasterIntent()`.

This should be enough.

Phew, I hope that is mostly clear. I was always afraid of explaining this because I know I'd definitely fail at it, shoot any questions you have.

It seems that ThreeJS is only dependent on Patchwork and not on Escher. But this PR needs both, so you can probably put these definitions in `@require Escher begin ... end`